Add a tiles_changed_parallel.

Elabajaba opened the pull request:
This is a bit of a widespread pr, so far I've tried to break up the different parts into separate commits, but I can submit some of them as separate pr's if you want.

All performance tests were done using a 1024x1024 random map with a 32 chunk size, and MooreNeighborhood unless otherwise specified.

Changes:

- Added tiles_changed_parallel (~280ms to ~80ms vs existing tiles_changed).
  - Currently only gets the paths for chunks in parallel (per-chunk parallelization).
- replaced uses of the std hashmap and hashset with hashbrown's for rayon support. (eg. par_keys)
- Fnv has been replaced with hashbrown's default aHash because I found it to be faster in my testing.
- PathCache chunks was reworked into a 1d vector for easier iteration. (was a 2d vector previously)
- Slight optimization to the existing tiles_changed (saved ~12ms)

mich101mich commented on Cargo.toml:
```diff
@@ -9,14 +9,28 @@ readme = "readme.md"
+profiling = { version="1.0.3", features=["profile-with-optick"] }
+criterion = "0.3.4"
+
+[[bench]]
+name = "my_benchmark"
+harness = false
```
"can't find `my_benchmark` bench"
If you have a good benchmark, please add it to the repo and add the folder to the "exclude" section in the Cargo.toml

mich101mich commented on Cargo.toml:
```diff
@@ -9,14 +9,28 @@ readme = "readme.md"
 [dependencies]
-fnv = "1.0"
-rayon = { version = "1.5", optional = true }
+hashbrown = { version="0.11.2", features=["rayon"] }
```
```suggestion
hashbrown = "0.11.2"
```
`features=["rayon"]` breaks code that doesn't want the rayon-feature. See features suggestion below for alternative

mich101mich commented on Cargo.toml:
```diff
@@ -9,14 +9,28 @@ readme = "readme.md"
+
 
 [features]
 default = ["rayon"]
```
```suggestion
default = ["parallel"]
parallel = ["rayon", "hashbrown/rayon"]
```
only enable hashbrowns rayon if our rayon is active (Note: requires renaming feature to `parallel` instead of `rayon`)

mich101mich commented on src/path_cache.rs:
```diff
@@ -26,21 +26,21 @@ where
+    ($obj: ident, $index: ident) => {
+        &mut $obj.chunks[$index]
     };
 }
```
`get_chunk_mut!(self, chunk_index)` is longer and less readable than `&mut self.chunks[chunk_index]`. These macros only existed for the point -> chunk_index conversion that is now in `get_chunk_index`. Please remove them.

Elabajaba replied:
Removed them.

I've currently got a mix of
```rust 
let chunk = self.get_chunks(cp)
// and
let chunk_index = self.get_chunk_index(cp);
let chunk =  &self.chunks[chunk_index];
```

depending on if I need to mutably borrow self. Should I keep using self.get_chunks() wherever possible and the 2nd one where needed, or should I just use the 2nd one everywhere?

mich101mich replied:
Seems fine the way it is.

mich101mich commented on src/path_cache.rs:
```diff
@@ -898,9 +900,11 @@ impl<N: Neighborhood + Sync> PathCache<N> {
+                .iter()
+                .filter(|&&pos| self.nodes.id_at(pos).is_none())
+                .map(|&a| a)
+                .collect();
```
```suggestion
            // Only include nodes that aren't already part of the map
            candidates.retain(|&pos| self.nodes.id_at(pos).is_none());
```

Elabajaba replied:
This is much cleaner, thanks!

mich101mich commented on src/path_cache.rs:
```diff
@@ -938,12 +945,248 @@ impl<N: Neighborhood + Sync> PathCache<N> {
+
+            for id in removed {
+                chunk.nodes.remove(&id);
+                self.nodes.remove_node(id); // This is where the majority of time is spent here
```
Suggestion: Collect all to-be-removed nodes from all chunks into a set and add a method called `remove_nodes` to NodeMap like this:
```rust
    #[track_caller]
    pub fn remove_nodes(&mut self, ids: &NodeIDSet) {
        for id in ids {
            let node = self.nodes[*id as usize].take().unwrap();
            self.pos_map.remove(&node.pos);
        }
        #[cfg(feature = "parallel")]
        {
            use rayon::prelude::*;
            self.nodes.par_iter_mut().for_each(|node| {
                if let Some(node) = node.as_mut() {
                    node.edges.retain(|id, _| !ids.contains(id));
                }
            });
        }
        #[cfg(not(feature = "parallel"))]
        self.nodes
            .iter_mut()
            .filter_map(|n| n.as_mut())
            .for_each(|node| {
                node.edges.retain(|id, _| !ids.contains(id));
            });
    }
```
I don't have your benchmarks so I cannot test the exact impact right now. Also please check if the `par_iter_mut` is an improvement or not.

mich101mich replied:
Nope, sorry. That option is way slower.

Elabajaba replied:
I managed to get a decent speedup here in bfedcb8aed708bae4124754239b07d12d6c845f5.

For whatever reason, it slows down the next section (recreates sides in renew) by about ~1ms, but it's still an overall gain of about 3ms for me.

Elabajaba replied:
Running into panics even after I fixed the original faulty logic issue, so I reverted it.

edit: Also improved the update_path test to catch it.

mich101mich commented on src/path_cache.rs:
```diff
@@ -938,12 +945,248 @@ impl<N: Neighborhood + Sync> PathCache<N> {
+
+        let size = self.config.chunk_size;
+
+        use std::time::Instant;
+        let outer_timer = Instant::now();
```
Please remove this once the discussion is over

Elabajaba replied:
Yup, that was just the easiest way I found to figure out where the expensive parts of the function were.

Elabajaba replied:
Now that we've migrated these to the logging crate and macros, I'm wondering if we should leave these in? (maybe behind eg. a perf-tracing flag?)

mich101mich replied:
Is now hidden behind a feature and `log` is optional. Run `cargo bench --features log` to get the full spam.

mich101mich commented on src/path_cache.rs:
```diff
@@ -938,12 +945,248 @@ impl<N: Neighborhood + Sync> PathCache<N> {
+
+        let mid_timer = Instant::now();
+
+        let paths: Vec<HashMap<u32, HashMap<(usize, usize), Path<(usize, usize)>>>> = {
```
```suggestion
        let paths: Vec<_> = {
```
or
`Vec<NodeIDMap<PointMap<Path<Point>>>>`

mich101mich commented on src/path_cache.rs:
```diff
@@ -938,12 +945,248 @@ impl<N: Neighborhood + Sync> PathCache<N> {
+
+        let inner_timer = Instant::now();
+
+        for thing in paths {
+            for (id, thing2) in thing {
```
Option 1:
- `paths` -> `all_paths` or leave as `paths`
- `thing` -> `paths` or `paths_in_chunk`
- `id` -> `src_id`
- `thing2` -> `paths_from_src`

Option 2:
Move part of this back into chunk and have `add_nodes_parallel` return a `Vec<(NodeId, NodeId, PathSegment)>` that directly contains the parameters to `self.nodes.add_edge`.
I would prefer Option 2.

mich101mich commented on src/path_cache.rs:
```diff
@@ -1014,7 +1257,20 @@ impl<N: Neighborhood + Sync> PathCache<N> {
+        let size = self.config.chunk_size;
+
+        let (x, y) = ((point.0 / size), (point.1 / size));
+        let chunk_count_x = match self.width % size {
+            0 => self.width / size,
+            _ => (self.width / size) + 1,
+        };
```
The same check happens in `new_internal`, so just add a member to PathCache to store this

mich101mich commented on src/path_cache.rs:
```diff
@@ -1263,6 +1519,97 @@ mod tests {
+        fn cost_fn(grid: &[[usize; 5]; 5]) -> impl '_ + Fn((usize, usize)) -> isize {
+            move |(x, y)| [1, 10, -1][grid[y][x]]
+        }
+        let pathfinding = PathCache::new_parallel(
```
```suggestion
        let pathfinding = PathCache::new(
```

mich101mich commented on src/path_cache.rs:
```diff
@@ -1263,6 +1519,97 @@ mod tests {
+        assert_eq!(pathfinding.get_chunk_index(point), 2);
+    }
+
+    #[test]
```
```suggestion
    #[test]
    #[cfg(feature = "parallel")]
```

mich101mich commented on src/path_cache.rs:
```diff
@@ -1263,6 +1519,97 @@ mod tests {
+    }
+
     #[allow(unused)]
     // #[test]
```
```suggestion
    // #[test]
    #[cfg(feature = "parallel")]
```

mich101mich reviewed (commented):
First of all thank you for making these changes!

Some general notes:
- The new tiles_changed_parallel is a lot of duplication. From what I see it seems like only the bottom part is different, so unless you plan on making more changes (I don't have your detailed benches, so please say if there are any bottlenecks there), I would ask you to please:
  - Add a method that contains the top part until the parallel segment OR
  - Add tiles_changed_internal that works similar to new_internal.
- I doubt that aHash is faster than Fnv based on their source code, but I could be wrong. I only compared Fnv to the std Hasher, and that was a big difference.
- Please bump the version number to 0.5.0, because the Fnv stuff is breaking changes

mich101mich commented on src/path_cache/chunk.rs:
```diff
@@ -208,6 +210,40 @@ impl Chunk {
         }
     }
 
+    pub fn add_nodes_parallel<N: Neighborhood + Sync>(
```
```suggestion
    #[cfg(feature = "parallel")]
    pub fn add_nodes_parallel<N: Neighborhood + Sync>(
```
To avoid "unused" warnings

Elabajaba commented:
The parallel features should all be resolved now.

Elabajaba commented:
I cleaned my benchmark up a bit and added it.

They're a bit bugged currently, as the create_pathcache benchmarks for large maps run quite a bit slower than running the same code as a separate binary on Windows (eg. ~3s vs ~800ms for a random map on my machine). They seem to run fine on WSL2 however.

The values I selected for map_size and chunk_size are arbitrary (tiny maps tend to  have a lot more variance between runs however), same with the start and end locations for the get path benchmarks.

To run a subset of the benchmarks, you can run it with `cargo bench -- BENCHNAME` (note the space between -- and the bench name), eg `cargo bench -- "Get Single Path, Large Random"`  to run a single benchmark, or `cargo bench -- "Create PathCache"` to run all the benchmarks in the Get Path group.

In order to get more specific timings of big functions (eg. tiles_changed), I just sprinkle in std::time::Instant::now() and output the timings in trace!() calls. Those get really spammy though, since I can't seem to find a way to limit the actual amount of runs criterion does, so I set the default env_logger output level to debug. You can change it to trace, but I wouldn't recommend doing it for more than 1 test at a time as it likes to spam out hundreds of lines.

The Get Path benchmarks for mid-sized maps (128x128) are pretty unstable in terms of timing, so I'm considering removing them (small 32x32 maps were even worse, with 50%+ variations being common between runs). The large 1024x1024 maps don't seem to have this issue.

mich101mich commented:
Now that I have run your Benchmark, I have to seriously ask what kind of machine you are running on to get these results...

My PC (Windows, 6 Threads) gets 220ms on both single thread and parallel, and my Laptop (Linux, 4 Threads) has 260ms single thread, 200ms parallel.

I have now further optimized connect_nodes (used to take 60ms, now 2ms) and turned the content of chunk.add_nodes to use par_iter as well, and now it is down to 50ms in the parallel case. You apparently went from 280ms to 80? how?

(btw I will commit those changes after merging this PR)

Elabajaba commented:
I've got a 16c/32t 3950x.

I didn't really consider what effect that would effect the parallel parts vs more standard 4-8 core machines...

mich101mich commented:
> * I doubt that aHash is faster than Fnv based on their source code, but I could be wrong. I only compared Fnv to the std Hasher, and that was a big difference.

I was actually able to confirm that hashbrowns default is _way_ better than Fnv, and it is actually better than the NodeIDHasher that I used (8% for 128x128 to 80% for 1024x1024 speedup in get_path)...
My best guess is that simpler Hashers have a better hashing time, but many many more hash collisions, slowing down the overall process at larger Map/Set sizes? Whatever. aHash it is.

Elabajaba commented:
aHash (hashbrown's default) is also just ridiculously fast. (see the [smhasher benchmarks](https://rurban.github.io/smhasher/doc/table.html))

Wyhash might be faster, but it has some issues with bad seeds (that can be mitigated at the language specific library level by just disallowing the bad seeds), and afaik doesn't implement the BuildHasher trait so it can't be used as a drop in replacement in hashmaps. The Wyhash2 crate does implement the BuildHasher trait, but it wasn't any faster than aHash in my testing, and I had to replace all the with_capacity declarations with with_capacity_and_hasher, and manually seed the hasher.

mich101mich commented:
Alright, it seems to me like we are pretty much done here. Is there anything that you want to add/that is missing/that doesn't work yet?

Elabajaba commented:
Looks good to me.
